File: 11-container-with-most-water/11-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int i = 0;
        int j = height.size() - 1;
        
        int maxArea = 0;
        
        while (i < j) {
            int curArea = min(height[i],height[j]) * (j-i);
            
            maxArea = max(maxArea,curArea);
            
            if(height[i] < height[j]) {
                i++;
            } else {
                j--;
            }
        }
        
        return maxArea;
    }
};

File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        left = 0
        right = len(height) - 1
        
        maxWater = float("-inf")
        
        while left < right:
            area = (right-left) * min(height[left],height[right])
            
            maxWater = max(maxWater,area)
            
            if height[left] < height[right]:
                left+=1
            else:
                right-=1
                
        return maxWater

# Time Complexity: O(n)
# Space Complexity: O(1)